<a href="https://colab.research.google.com/github/Rafaelcarvalho320/Homework5-e-Homework_6b_SVM_AdaBoost/blob/main/Homework5_reso.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Homework_5_SVMs**

Arienilce Sacramento - 201706840081

Rafael Carvalho - 201706840055

Yasmim Miranda - 201906840010


Dataset usado: estudante_201706840055

6) Use o scikit-learn para treinar uma SVM com kernel linear, usando hiperparâmetro: C=1.

Neste caso use a versão normalizada dos dados, com os fatores de normalização projetados com base no conjunto de treino, para que este conjunto tenha média 0 e variância 1.

a) Treine a SVM com o conjunto de treino indicado no arquivo dataset_train.txt.
b) Indique o desempenho do modelo SVM no conjunto de teste dataset_test.txt. 


In [ ]:
from sklearn import datasets
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np


In [ ]:
# Carregar o conjunto de treinamento
df_train = pd.read_csv(r'C:\Users\Dell\Documents\SIGAA\8º Semestre\Inteligência Computacional\estudante_201706840055_train.txt')


# Converter o dataframe do pandas para uma matriz NumPy
data_train = np.array(df_train)
X_train = data_train[:, :-1]  
y_train = data_train[:, -1]   

# Carregar o conjunto de teste
df_test = pd.read_csv(r'C:\Users\Dell\Documents\SIGAA\8º Semestre\Inteligência Computacional\estudante_201706840055_test.txt')

# Converter o dataframe do pandas para uma matriz NumPy
data_test = np.array(df_test)
X_test = data_test[:, :-1]  
y_test = data_test[:, -1]   

In [ ]:
sc = StandardScaler()
X_train = sc.fit_transform(X_train)


In [ ]:
svm_linear = SVC(kernel='linear', C=1)


In [ ]:
svm_linear.fit(X_train, y_train)


SVC(C=1, kernel='linear')

In [ ]:
X_test = sc.transform(X_test)


In [ ]:
y_pred = svm_linear.predict(X_test)


In [ ]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.33      1.00      0.50         1
           1       1.00      0.33      0.50         3

    accuracy                           0.50         4
   macro avg       0.67      0.67      0.50         4
weighted avg       0.83      0.50      0.50         4



c) Converta esta SVM para um perceptron, estime o custo computacional durante a fase de teste da SVM na forma original com sua versão implementada como perceptron. Indique se há algum ganho computacional em termos de memória e número de operações ao se converter essa SVM para um perceptron.

In [ ]:
from sklearn import datasets
from sklearn.linear_model import Perceptron
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np


df_train = pd.read_csv(r'C:\Users\Dell\Documents\SIGAA\8º Semestre\Inteligência Computacional\estudante_201706840055_train.txt')


data_train = np.array(df_train)
X_train = data_train[:, :-1] # Todas as colunas, exceto a última
y_train = data_train[:, -1] # Última coluna, que contém os rótulos/classes


df_test = pd.read_csv(r'C:\Users\Dell\Documents\SIGAA\8º Semestre\Inteligência Computacional\estudante_201706840055_test.txt')


data_test = np.array(df_test)
X_test = data_test[:, :-1] # Todas as colunas, exceto a última
y_test = data_test[:, -1] # Última coluna, que contém os rótulos/classes

sc = StandardScaler()
X_train = sc.fit_transform(X_train)

perceptron = Perceptron()

perceptron.fit(X_train, y_train)

X_test = sc.transform(X_test)

y_pred = perceptron.predict(X_test)

from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.33      1.00      0.50         1
           1       1.00      0.33      0.50         3

    accuracy                           0.50         4
   macro avg       0.67      0.67      0.50         4
weighted avg       0.83      0.50      0.50         4



O custo computacional durante a fase de teste da SVM depende do número de vetores de suporte e da dimensão do espaço de característica. Já a versão implementada como perceptron tem um custo computacional durante a fase de teste de O(d), independentemente do número de exemplos de treinamento ou do número de vetores de suporte.
Em termos de memória, a SVM armazena todos os vetores de suporte, que podem ser um conjunto substancial de dados, enquanto o perceptron armazena apenas os pesos treinados, que geralmente são muito menores do que os vetores de suporte.

7) Sobre o projeto de SVMs: 
a) Use o scikit-learn (exemplificado no código ak_svm.py) para treinar uma SVM com kernel Gaussiano (também chamado de “RBF”), usando o conjunto de validação para fazer o “tuning” (otimização) de dois hiperparâmetros: “C” e “gamma”, avaliando os valores: 

C = 0.01
C = 1
C = 100 

gamma = 0.5
gamma = 1

Em suma:
1) Treine as SVMs com o conjunto de treino variando C e gamma;
2) Observe o resultado usando o conjunto de validação;
3) Indique qual SVM apresenta melhor performance no conjunto de validação. 

Obs: O produto Cartesiano das opções de hiperparâmetros irá exigir 3 x 2 = 6 treinos de SVMs.  


In [ ]:
import pandas as pd
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

# Carregar o conjunto de treinamento e validação
train_data = pd.read_csv(r'C:\Users\Dell\Documents\SIGAA\8º Semestre\Inteligência Computacional\estudante_201706840055_train.txt')

val_data = pd.read_csv(r'C:\Users\Dell\Documents\SIGAA\8º Semestre\Inteligência Computacional\estudante_201706840055_validation.txt')

data_train = np.array(df_train)
X_train = data_train[:, :-1] 
y_train = data_train[:, -1] 

data_val = np.array(df_train)
X_val = data_train[:, :-1] 
y_val = data_train[:, -1] 

# Normaliza os dados
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_val = sc.transform(X_val)

# Valores de C e gamma 
param_grid = {'C': [0.01, 1, 100], 'gamma': [0.5, 1]}

# Loop para treinar as SVMs com os diferentes valores de C e gamma e avaliar a performance no conjunto de validação
best_accuracy = 0
for C in param_grid['C']:
    for gamma in param_grid['gamma']:
        svm = SVC(kernel='rbf', C=C, gamma=gamma)
        svm.fit(X_train, y_train)
        y_pred = svm.predict(X_val)
        accuracy = accuracy_score(y_val, y_pred)
        print(f'SVM com C={C}, gamma={gamma}: validation accuracy = {accuracy}')
        if accuracy > best_accuracy:
            best_accuracy = accuracy
            best_svm = svm

print(f'Nº de vetores de suporte: {best_svm.n_support_}')
# Obter os índices das amostras de treino selecionadas como vetores de suporte
support_indices = best_svm.support_

# Obter os valores dos "lambdas" e o termo independente b 
lambdas = best_svm.dual_coef_
bias = best_svm.intercept_

print(f'Índices das amostras de treino selecionadas como vetores de suporte: {support_indices}')
print(f'Valores dos "lambdas": {lambdas}')
print(f'Termo independente b: {bias}')

# Imprimir a SVM com melhor performance
print(f'Melhor SVM: {best_svm} com accuracy = {best_accuracy}')



SVM com C=0.01, gamma=0.5: validation accuracy = 0.5714285714285714
SVM com C=0.01, gamma=1: validation accuracy = 0.5714285714285714
SVM com C=1, gamma=0.5: validation accuracy = 0.5714285714285714
SVM com C=1, gamma=1: validation accuracy = 1.0
SVM com C=100, gamma=0.5: validation accuracy = 1.0
SVM com C=100, gamma=1: validation accuracy = 1.0
Nº de vetores de suporte: [3 4]
Índices das amostras de treino selecionadas como vetores de suporte: [0 1 2 3 4 5 6]
Valores dos "lambdas": [[-1.         -1.         -1.          0.13719668  0.86280332  1.
   1.        ]]
Termo independente b: [0.35086436]
Melhor SVM: SVC(C=1, gamma=1) com accuracy = 1.0


b) Descreva totalmente (por extenso) os parâmetros que compõem a fórmula final do classificador SVM escolhido, indicando:

b1) Os parâmetros escolhidos do kernel (gamma neste caso, visto que C é apenas um hiperparâmetro, usado no treino mas não no teste) 

b2) O número de vetores de suporte (SVs)

b3) Seus índices no conjunto de treino

b4) Os respectivos valores dos “lambdas” (ou “dual_coef_” de coeficientes duais) e 

b5) O termo independente b chamado de “bias” ou “intercept_”.

Resposta:
b1) O parâmetro escolhido do kernel é gamma, com valor igual a 1.
b2) O número de vetores de suporte é composto por 3 amostras da classe negativa e 4 amostras da classe positiva.
b3) Os índices das amostras de treino selecionadas como vetores de suporte são: [0, 1, 2, 3, 4, 5, 6].

b4) Os valores dos “lambdas” (ou “dual_coef_” de coeficientes duais) são: [[-1. -1. -1. 0.13719668 0.86280332 1. 1.]].

b5) O termo independente b é de 0.35086436.

1) A figura abaixo mostra os exemplos de treino e as regiões de decisão de uma SVM (ou SVC, de “support vector classifier”) para um problema binário (duas classes) onde o vetor x de “features” de entrada tem dois parâmetros (weight e length). Os pontos em azul e vermelho, representam os exemplos de cada classe. As regiões em azul e vermelho representam as regiões nas quais o classificador irá prever uma classe azul e vermelha, respectivamente. Olhando para as figuras, informe: 

a) Quantos erros cada SVM possui no conjunto de treino ? 
b) Qual a que lhe parece ser a melhor SVM? Por que?


Resposta:
a) Com excessão da SVC with polynomial kernel,que não há erro, e da LinearSVC, com 2 erros, todas possuem 1 erro no conjunto de treino.
b) A melhor parece ser a SVC with linear kernel. Pois nela as regiões de decisão estão bem afastadas umas das outras. 

2) A regularização é uma importante ferramenta para controlar a “complexidade do modelo” durante a fase de seu projeto. No projeto de uma SVM, o hiperparâmetro C é usado para efetuar a regularização do modelo na etapa de seleção do modelo, ou seja, quando os hiperparâmetros são variados e observamos o impacto através do erro de classificação no conjunto de validação. Suponha que você se encontra em uma situação onde está descontente com o atual resultado de sua SVM e observa que o número de vetores de suporte está relativamente pequeno. Você quer potencialmente aumentar o número de vetores de suporte, fazendo com que sua nova SVM seja um modelo mais “complexo”. Para isso, você deve aumentar ou diminuir o parâmetro “C” da classe SVC do scikit-learn?

Para aumentar o número de vetores de suporte, o parâmetro C deve ser diminuído. Isso ocorre porque o parâmetro C controla a margem da SVM, e quanto maior o valor de C, menor será a margem e maior será a penalidade por classificações incorretas

3) Um classificador usa vetores de entrada com dimensão K=5 “features”. Após treinar uma SVM linear, o número de vetores de suporte foi de 450 exemplos. Neste caso, calcular o kernel linear corresponde a um produto interno entre dois vetores de dimensão K=5, cada, o que requer K multiplicações e K-1 adições. Estime o fator F = Coriginal / Cperceptron de redução do custo computacional da etapa de teste ao converter esta SVM linear para um perceptron. Assuma que os custos Coriginal e Cperceptron correspondem ao número de multiplicações e adições usando-se, respectivamente, a SVM original com os 450 vetores de suporte e após sua conversão para perceptron.

Para a SVM linear, precisamos calcular o produto interno entre o vetor de entrada e cada um dos 450 vetores de suporte. Isso requer 450 operações de produto escalar, cada um dos quais envolve 5 multiplicações e 4 adições, totalizando 

C(original) = 450 * 5 * 4 = 9.000 operações de multiplicação e adição.

Para o perceptron, precisamos calcular o produto interno entre o vetor de entrada e o vetor de pesos. Isso requer 5 multiplicações e 4 adições, totalizando

C(perceptron) = 5 * 4 = 20 operações de multiplicação e adição.

F = C(original) / C(perceptron) = (9.000 operações) / (20 operações) ≈ 450

Da SVM linear para um perceptron reduz o custo computacional de avaliação por um fator de aproximadamente 450

4) Interpretação do resultado de projeto de SVMs.
O treinamento de uma SVM linear usando a classe SVC do sklearn retornou o resultado:
svm.n_support_= [1 2]  # número de vetores de suporte por classe
svm.support_vectors_= 	[[ 1. 4.]  # vetores de suporte
 [-2.  3.]
 [-2. -5.]]
svm.dual_coef_= [[-0.5 -0.3 0.8]] # valores de \lambda
svc.intercept_= [-2] # “bias”
Observe que esta questão não usa o conjunto de treino indicado, mas outro. Pede-se:
a) A função de decisão para essa SVM de acordo com a fórmula geral:


fórmula geral:

f(z) = (Σ lambda_i * K(z,x_i)) + b
Substituindo os valores dados na questão:

f(z) = (-0.5K(z,[1,4])) + (-0.3K(z,[-2,3])) + (0.8*K(z,[-2,-5])) - 2

b) A função de decisão desta SVM quando escrita como um perceptron f(z)=<z,w>+b.
c) A saída da função de decisão quando o vetor de entrada é z=[0, 0] e o respectivo rótulo predito y assumindo-se que y = I( f(z) > 0 ), onde I(.) é a função “indicador”, que é 1 se o argumento é verdadeiro, ou 0 em caso contrário.


b)
f(z) = <z, w> + b
Primeiro, é preciso calcular os valores de w multiplicando os valores dos vetores de suporte pelos valores de lambda:
w = Σ lambda_i * x_i

w = (-0.5 * [1, 4]) + (-0.3 * [-2, 3]) + (0.8 * [-2, -5])
w = [1.1, 2.9]
b= -2

função de decisão da SVM quando escrita como um perceptron é:
f(z) = <z, [1.1, 2.9]> - 2


c)
Substituindo o valor de z=[0,0]

f([0,0]) = (0.5 * (01 + 04)) + (-0.3 * (0*(-2) + 03)) + (0.8 * (0(-2) + 0*(-5))) - 2 
= -2
Como f([0,0]) não é maior do que zero, temos que y = I(f([0,0])) = 0. Portanto, o rótulo predito para o vetor de entrada z=[0,0] é 0.

5) Interpretação do resultado de projeto de SVMs.
O código ak_svm_prova1.py e o conjunto de treino em dataset_train.txt foram usados para gerar a figura e o log (saída de texto) abaixo. Em suma, foram treinadas 4 SVMs, duas com kernel linear (SVMs 1 e 2) e outras duas com kernel não-linear (SVMs 3 e 4). 
Pergunta-se:
a) Quais as expressões para as SVMs 3 e 4 (não-lineares) e para a SVM 2 (linear) de acordo com a nomenclatura abaixo (escreva a “fórmula” da função de decisão de cada SVM indicando o valor do bias “b”, etc., considerando que z e xn são vetores já normalizados (com o primeiro elemento dividido por 1000).



Para a SVM 2

f(z) = (-0.68611287 * z[0] - 0.10653139 * z[1]) - 0.99556119

Para a SVM 3 (RBF)

f(z) = (-0.91722233 * K(z,x0) - 0.91351914 * K(z,x1) - 0.91300432 * K(z,x2) + 0.87185969 * K(z,x3) + 0.8718861 * K(z,x4) + 1 * K(z,x5)) - 0.08676121

Para a SVM 4

f(z) = (-0.00887134 * K(z,x0) - 0.03133903 * K(z,x1) + 0.04021037 * K(z,x2)) - 1.03731897

Para SVM 1
f(z) = <z, w> + b = (-0.686) * (z[0]/1000) + (-0.107) * (z[1]/1000) - 0.996

Para SVM 2
f(z) = <z, w> + b = (-1.200) * (z[0]/1000) + (-0.200) * (z[1]/1000) - 1.800

c) Suponha que para a SVM 2 (linear), apenas as informações abaixo fossem fornecidas. Com elas, você já pôde escrever essa SVM no formato geral de uma SVM, como no item a). Explique agora com clareza quais os passos que você adotaria para converter essa SVM linear em um perceptron como descrito no item b) e indique em termos do número de multiplicações e adições, qual economia no custo computacional que a implementação como perceptron alcança.

Para converter a SVM linear em um perceptron, é necessário obter os valores dos pesos (w)
w = Σ lambda_i * x_i

w = (-0.45994152 * [0, -4]) + (-0.27992202 * [-1, 2]) + (0.73986354 * [-2, -2]) = [1.06045492, -3.52044158]
b= -1,79

f(z) = <z, [1.06045492, -3.52044158]> - 1.79

d) Para as SVMs 3 e 4 (não-lineares), indique:
d.1) qual o número total de vetores de suporte (SVs), considerando todas as classes.
d.2) Quais os índices desses SVs no conjunto de treino e os respectivos valores dos “lambdas” (coeficientes duais)
d.3) Qual o valor do termo independente b chamado de “bias” ou “intercept_”


Para a SVM 3:
d1) O número total de vetores de suporte é 6.
d2) Os índices dos SVs são [0, 1, 2, 3, 4, 5] e os respectivos valores de lambdas são [-0.91722233, -0.91351914, -0.91300432, 0.87185969, 0.8718861, 1.0].
d3) O valor de b é -0.08676121.

Para a SVM 4:
d1) O número total de vetores de suporte é 3 (2 para a primeira classe e 1 para a segunda classe).
d2) Os índices dos SVs são [0, 1, 5] e os respectivos valores de lambdas são [-0.00887134, -0.03133903, 0.04021037].
d3) O valor de b é -1.03731897.

e) Considerando as saídas da SVM abaixo para o conjunto de treino, em qual dos exemplos de treino esta SVM está menos “confiante” (assumindo que esses números são “confidence scores”) em sua decisão e qual é a classe que esta SVM prediz para este exemplo?
svm.decision_function(X)= [-1.00027976 -1.00027976 -0.99977173  1.00010297  1.00022828  0.90993821]


A SVM está menos "confiante" em sua decisão no terceiro exemplo de treino, que recebeu um score de decisão de -0.99977173.
A SVM prediz a classe  para o quinto exemplo de treino